Use metrics handles in transforms

conorbros opened the pull request:
- Using metrics handles in transforms is pretty simple. 
- Using handles instead of the macros in `TransformChain` and `Wrapper` will require more complex changes. Will see if it's worth re-architecting things to do this.

rukai commented on shotover-proxy/src/server.rs:
```diff
@@ -99,6 +99,9 @@ pub struct TcpCodecListener<C: Codec> {
     message_count: u64,
+
+    /// Metric recording shotover available connections
+    gauge: Gauge,
```
Can we name the metrics properly rather than putting the name in a comment?
Something like: `available_connections_gauge: Gauge`

rukai reviewed (approved):
Improved type safety and performance, nice!

Happy to merge as is but some thoughts:
There are some cases where second handle is created with the same args as another handle, will that result in duplicate metrics entries? And if yes, is that undesirable?

I think the TransformChainMetrics struct doesnt really do much for us.
It does have the advantage that we dont have to write `#[derivative(Debug = "ignore")]` twice but I think we would just be better off with less indirection.

conorbros commented:
There doesn't appear to be any side effects re-registering a metric. Also I presume once we implement a solution for https://github.com/shotover/shotover-proxy/issues/337 we will use the transform names in the metrics so it won't be the case for long.

I think that's reasonable, I'll remove the `TransformChainMetrics` struct.

Yeah I think it can be closed given we are using metric handles wherever possible

conorbros commented:
Ok, I've addressed the feedback. 

I've realized that the `new_no_shared_state` function does nothing different to the `new` function? I will make a PR after this to remove it.

rukai reviewed (approved):
passed x2

benbromhead reviewed (approved):
It looks like ithis changes some of the metrics names

rukai commented:
oh, I missed that, we should avoid that where possible

conorbros commented:
Yep that wasn't the intention. Which one changed?

benbromhead commented:
> Yep that wasn't the intention. Which one changed?
> 
> Edit: @benbromhead Are you referring to the ones with the transform names? e.g.
> 
> ```rust
> let failed_requests = register_counter!("failed_requests", "chain" => chain_name.clone(), "transform" => "CassandraSinkSingle");
> ```
> 
> Those haven't changed, I just removed the method that returned the transform name and put it as a literal in the metric.

Ah gotcha... all g if nothing has changed
